core/src/smartcontracts/isi/permissions.rs

```diff
@@ -20,65 +23,362 @@ impl NeedsPermission for QueryBox {}
+    World(#[skip_container] Box<dyn IsAllowed<World, QueryBox> + Send + Sync>),
+    /// Validator for [`MockWorld`]
+    #[cfg(test)]
+    Mock(#[skip_container] Box<dyn IsAllowed<MockWorld, QueryBox> + Send + Sync>),
+}
 
-/// Box with permissions validator for `Query`.
-pub type IsQueryAllowedBoxed<W> = IsAllowedBoxed<W, QueryBox>;
+#[allow(clippy::panic_in_result_fn, clippy::unimplemented)]
+impl IsAllowed<World, QueryBox> for IsQueryAllowedBoxed {
+    fn check(
+        &self,
+        authority: &AccountId,
+        operation: &QueryBox,
+        wsv: &WorldStateView<World>,
+    ) -> Result<()> {
+        match self {
+            IsQueryAllowedBoxed::World(query) => query.check(authority, operation, wsv),
+            #[cfg(test)]
+            IsQueryAllowedBoxed::Mock(_) => unimplemented!(),
+        }
+    }
+}
+
+/// Box with permissions validator for [`Expression`].
+#[derive(Debug, Clone, FromVariant, Serialize)]
+pub enum IsExpressionAllowedBoxed {
+    /// Validator for [`World`]
+    World(#[skip_container] Box<dyn IsAllowed<World, Expression> + Send + Sync>),
+    /// Validator for [`MockWorld`]
+    #[cfg(test)]
+    Mock(#[skip_container] Box<dyn IsAllowed<MockWorld, Expression> + Send + Sync>),
+}
+
+#[allow(clippy::panic_in_result_fn, clippy::unimplemented)]
+impl IsAllowed<World, Expression> for IsExpressionAllowedBoxed {
+    fn check(
+        &self,
+        authority: &AccountId,
+        operation: &Expression,
+        wsv: &WorldStateView<World>,
+    ) -> Result<()> {
+        match self {
+            IsExpressionAllowedBoxed::World(expression) => {
+                expression.check(authority, operation, wsv)
+            }
+            #[cfg(test)]
+            IsExpressionAllowedBoxed::Mock(_) => unimplemented!(),
+        }
+    }
+}
 
 /// Trait for joining validators with `or` method, auto-implemented
-/// for all types which convert to [`IsAllowedBoxed`].
-pub trait ValidatorApplyOr<W: WorldTrait, O: NeedsPermission> {
+/// for all types which are convertible to something implementing [`IsAllowed`]
+pub trait ValidatorApplyOr<W: WorldTrait, O: NeedsPermission, V: IsAllowed<W, O>>: Into<V> {
     /// Combines two validators into [`Or`].
-    fn or(self, another: impl Into<IsAllowedBoxed<W, O>>) -> Or<W, O>;
+    ///
+    /// # Errors
+    /// If validators have different types
+    fn or(self, another: impl Into<V>) -> Or<W, O, V>;
 }
 
-impl<W: WorldTrait, O: NeedsPermission, V: Into<IsAllowedBoxed<W, O>>> ValidatorApplyOr<W, O>
-    for V
+impl<W: WorldTrait, O: NeedsPermission, V: IsAllowed<W, O>, I: Into<V>> ValidatorApplyOr<W, O, V>
+    for I
 {
-    fn or(self, another: impl Into<IsAllowedBoxed<W, O>>) -> Or<W, O> {
+    fn or(self, another: impl Into<V>) -> Or<W, O, V> {
+        Or::new(self, another)
+    }
+}
+
+/// `check` succeeds if either `first` or `second` validator succeeds.
+#[derive(Debug, Clone, Serialize)]
+pub struct Or<W: WorldTrait, O: NeedsPermission, V: IsAllowed<W, O>> {
```

Comment:
These will need to be replaced by generic, general-purpose combinators. We should start work on them after the LTS release.